Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/calculation/CheckInWarningMatcher.kt

```diff
@@ -0,0 +1,144 @@
+            Timber.e("Calculation partially failed.")
+            false
+        } else {
+            Timber.d("Matching was successful.")
+            true
+        }
+
+        return Result(
+            successful = successful,
+            processedPackages = matchLists.filterNotNull().flatten()
+        )
+    }
+
+    data class Result(
+        val successful: Boolean,
+        val processedPackages: Collection<MatchesPerPackage> = emptyList(),
+    )
+
+    @VisibleForTesting(otherwise = PRIVATE)
+    internal suspend fun createMatchingLaunchers(
+        checkIns: List<CheckIn>,
+        warningPackages: List<TraceWarningPackage>,
+        coroutineContext: CoroutineContext
+    ): List<Deferred<List<MatchesPerPackage>?>> {
+
+        val launcher: CoroutineScope.(
+            List<CheckIn>,
+            List<TraceWarningPackage>
+        ) -> Deferred<List<MatchesPerPackage>?> = { list, packageChunk ->
+            async {
+                try {
+                    packageChunk.flatMap {
+                        findMatches(list, it)
+                        packageChunk.map { warningPackage ->
+                            val overlaps = findMatches(list, warningPackage)
+                            MatchesPerPackage(warningPackage = it, overlaps = overlaps)
+                        }
+                    }
+                } catch (e: Throwable) {
+                    Timber.e(e, "Failed to process packages $packageChunk")
+                    null
+                }
+            }
+        }
+
+        // at most 4 parallel processes
+        val chunkSize = (checkIns.size / 4) + 1
+
+        return warningPackages.chunked(chunkSize).map { packageChunk ->
+            withContext(context = coroutineContext) {
+                launcher(checkIns, packageChunk)
+            }
+        }
+    }
+
+    data class MatchesPerPackage(
+        val warningPackage: TraceWarningPackage,
+        val overlaps: List<CheckInWarningOverlap>,
+    )
+}
+
+@VisibleForTesting(otherwise = PRIVATE)
+internal suspend fun findMatches(
+    checkIns: List<CheckIn>,
+    warningPackage: TraceWarningPackage
+): List<CheckInWarningOverlap> {
+    return warningPackage
+        .extractWarnings()
+        .flatMap { warning ->
+            checkIns
+                .mapNotNull { checkIn ->
+                    checkIn.calculateOverlap(warning, warningPackage.packageId).also { overlap ->
+                        if (overlap == null) {
+                            Timber.v("No match/overlap found for $checkIn and $warning")
+                        } else {
+                            Timber.i("Overlap found $overlap")
+                        }
+                    }
+                }
+        }
+}
+
+@VisibleForTesting(otherwise = PRIVATE)
+internal fun CheckIn.calculateOverlap(
+    warning: TraceWarning.TraceTimeIntervalWarning,
+    traceWarningPackageId: String
+): CheckInWarningOverlap? {
+
+    // TODO this is not correct anymore
```

Comment:
if its not correct anymore... whats the goal for this issue (new PR, fixing it here?)

Reply:
Not my comment but will see if we can resolve this here.

Reply:
Removed the TODO here and aligned the locationIdHash comparison:

https://github.com/corona-warn-app/cwa-app-android/pull/2707/commits/241e4a8e850fedc6144cf5db0b43362141079c78

I currently can't get any matches though, will have to re-check tomorrow.

Reply:
I can produce locationIdHash matches now against the mock server. :+1: